### ISSUE TITLE
Revisions should be the same type as what they are revising

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql#974]
Currently, Revisions in the Schema are treated as their own `Revision` Type. However, in real use cases, Revisions are the same Type (page, post, etc) as what they're revising. Their Type in the Schema should be the same. 

For example, this kind of query should work: 

```
{
  posts {
    nodes {
      id
      title
      someCustomFieldOnlyForPosts
      revisions {
          id
          title
          someCustomFieldOnlyForPosts
       }
    }
  }
}
```

We should be able to register things like custom fields to a Post Type, and revisions of that Post Type should expose those fields as well. 

A revision of a Post is still a Post. A revision of a Page is still a page. 

So, what I propose needs to change, is that all fields exposing revisions resolve to the same Type as the parent (for example `Page.revisions` should resolve to a `Page->Page` connection) and the `RootQuery.revisions` connection should resolve to a `PostObjectUnion` where the node can be one of _many_ Types. 

---

Additionally, a user should be able to fetch a single `page` or `post` (or CPT) using an ID of a revision. 

For example, if we had a revision of a Post with ID `5`, I should be able to do something like: 

```
{
  postBy( postId: 5 ) {
     id
     title
  }
}
```

If the ID is of a revision and the revision parent is a `Post`, then the revision should be resolved to the user, if the user has proper capabilities to view revisions.

**Comment by CodeProKid** [comment]
I'm not sure I agree with this. WordPress treats revisions as their own type by designating it as a post within the `revision` post_type. If it were meant to be of the same type as it's parent `revision` would be a post_status, not a post_type.

Also, WP core currently doesn't track any meta in revisions so I'm not sure that I see the upside of having the data the same shape, because any custom fields would be empty values. I'm also not sure that it tracks term relationships either. As far as I know it's only the core post object fields. So the only case where I actually see this being useful is if a post type is using the core fields for non-traditional data storage where they've mapped a core field to something else. Something like redirects where the `post_title` is actually the from url, and the `post_content` is the to url. That's the only use case where I can think of the data shape being the same actually being an advantage.

**Comment by jasonbahl** [comment]
@CodeProKid there are a lot of decisions WordPress made years ago that I think would have gone a different way if it were done today 😄 

From a declarative non-implementation standpoint, when looking at something like a preview of a Page, you're still looking at a Page. It's a revision of a Page, but it's still a Page. If you're looking at a revision of a Car, it's a Car. It's not a completely different Type of thing. 

At the moment, the `Revision` Type supports the "core" fields of a post_type, such as `title`, `date`, etc. . .

But many Post Types have different uses for these fields, and may even chose to rename them. 

For example, if I had a "Car" post type, I might rename "Title" to "Model" and in GraphQL it would be good to be able to ask for the `Model` of the revision instead of the `Car.model and Revision.title`

Right now, `Revision` is a single Type, so if you did want to revise meta (plugins like ACF do this), how would you even go about exposing the revised meta if `Revision` could be a Car, House, Post or Page?

I think having `fields` to ask for Revisions still make sense, but the shape should inherit from the Type they are revisions of.

**Comment by CodeProKid** [comment]
yeah, I see what you're getting at. My point is that if the the data that makes a `Car` a `Car` are entirely stored in meta, then the revision isn't a `Car` unless that meta data is stored on the revision, which is a big IF right now in WP.

My concern is just that we will be relying on third party tooling to sync the meta data over for the matching Type definition to make sense. So this makes a lot of sense for a developer using ACF, but what about someone using CMB2, metabox.io, or a custom build implementation?

I think we just need to be sure we aren't building against the assumption that a developer is using ACF (which does store meta on revisions).

**Comment by jasonbahl** [comment]
Ya, it actually didn't even come up because of ACF. 

It came up when we were thinking through how to build a UI for previewing a Post (or page or CPT) and it became clear that a Preview of a Post object is a revision of said object, and should have the same Schema shape as that object. 

Of course, plugins like ACF have already done work to play nice here. If I'm not mistaken I think @jaredcobb did some work to make Fieldmanager data also be revised as well. 

In any case, the `Revision` Type itself has very little functional utility that wouldn't be improved by changing revision field(s) to return the Types they are revisions of.

**Comment by CodeProKid** [comment]
I definitely agree that a single `Revision` does not work, as the shape will need to be different per post type with most custom implementations.

I'm still not totally convinced that it should be the _same_ type as it's parent, since that relies on a lot of assumptions for the underlying data to be the same (which I think needs to happen for it to make sense to use the same type). 

I wonder if a good middle ground would be something like `{$post_type}Revision` as the type. So for posts you would have a `PostRevision` type, and then plugins like WPGraphQL ACF could add the same fields to that type that it does to the parent type.

**Comment by jasonbahl** [comment]
I think a separate Type still is a disservice if you consider the _actual_ tangible use of revisions. 

Revisions, if you eliminate underlying implementation details, are intended to be an object over time. 

If you were to literally freeze a Post in time, that's what a revision would be. The WP core implementation lacks some implementation details, such as versioning meta, but from a declarative Schema perspective, a revision of a Post is a Post. A revision of a Page is a Page. 

If I paint my house, it's still a House, it just has a different value for `color`. My house wouldn't become a `HouseRevision` it would still be a House, but some properties would change over time. 

For building UIs that make use of viewing an object over time, having them be the same Type will have many tangible benefits, and I don't really think many down sides. 🤔

**Comment by CodeProKid** [comment]
I thought about this a lot, and I'm starting to warm up to this. I think there still needs to be more discussion / thought on how to use this for preview, since I don't think it's simple to do out of the box. I'd love to see a POC for how you imagine that working, I think that would clear up a lot of details for this implementation.

I do think we should add an `isRevision` bool field on any type that supports revision to make checking against this straight forward. It should maybe even be on all Types that follow the Post Type interface.

**Comment by jasonbahl** [comment]
- remove the RootQuery.revision & RootQuery.revisionBy
- remove User.revisions
- change RootQuery.revisions to return a connection to ContentRevisionUnion
- change {PostType}.revision to return a connection to a {$PostQuery}

**Comment by jasonbahl** [comment]
closed by #1056
